### DISCUSSION TITLE
Suggestion: Don't allow more than 1 consecutive blank line

### DISCUSSION THREAD

**Discussion by nfcampos** [tconbeer/sqlfmt#230] [The sqlfmt Style] [answered]
**Describe the bug**
Currently whitespace around CTEs is preserved sometimes, which for me again isn't desired, but I understand if this is intended and not something you want to change

**To Reproduce**
```
with






    line_items as (





        select * from {{ ref("order_line_items__01_with_financials") }}
        
    ),









    gift_cards as (
        
        
        
        select * from {{ ref("gift_cards") }}
    
    
    
    ),







select *
from line_items





union all





select *
from gift_cards


```

**Expected behavior**
ie. when new lines exist around CTEs / other things (see the space before `union all`) then I expect it to be reduced to a single new line. (I personally would even go further and add those single new lines even if they weren't in the input, but I do understand that's probably a minority opinion)
```
with

    line_items as (

        select * from {{ ref("order_line_items__01_with_financials") }}

    ),

    gift_cards as (select * from {{ ref("gift_cards") }}),

select *
from line_items

union all

select *
from gift_cards

```

**Actual behavior**
```
with
    line_items as (





        select * from {{ ref("order_line_items__01_with_financials") }}

    ),









    gift_cards as (select * from {{ ref("gift_cards") }}),







select *
from line_items





union all
select *
from gift_cards

```

**Additional context**
What is the output of `sqlfmt --version`?
```sqlfmt, version 0.10.0```

**Comment by tconbeer** [comment]
Right now we try to leave all blank lines as-is, unless we merge across them. There is a bug I introduced somewhere between 0.8 and 0.10 that eliminates some blank lines for reasons I don't understand (I need to make an issue for that).

I'm not totally comfortable with enforcing a style that is "you can never have more than one consecutive blank line," although in practice I don't think I write queries with more than one blank line in them.

What rule(s) around blank lines should we try to enforce?

**Comment by nfcampos** [comment, marked as answer]
The opinion of some formatters for other languages 
- https://prettier.io/docs/en/rationale.html#empty-lines
- https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#empty-lines

Both leave empty lines from original format, but collapse them to a standard number of lines (ie. 0 empty lines remain 0, 7 or 6 or 1 empty lines all become eg 1 empty line)

Black goes even further and adds empty lines in some places (which I think you could consider doing eg around CTEs for readability)
